File: packages/typescript/src/exporter/retry.ts
/**
 * Exponential backoff retry helper.
 *
 * Mirrors agent_capture.exporter.retry. Used by HTTPExporter for transient
 * failures. Permanent failures (4xx) skip retries via the `retryable`
 * predicate.
 */

export interface RetryPolicy {
  maxAttempts: number;
  baseDelayMs: number;
  maxDelayMs: number;
  multiplier: number;
  jitter: number;
  /** Override for tests. */
  sleep?: (ms: number) => Promise<void>;
}

export const defaultRetryPolicy: RetryPolicy = {
  maxAttempts: 5,
  baseDelayMs: 500,
  maxDelayMs: 30_000,
  multiplier: 2.0,
  jitter: 0.25,
};

function defaultSleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function delayForAttempt(policy: RetryPolicy, attempt: number): number {
  const raw = Math.min(
    policy.maxDelayMs,
    policy.baseDelayMs * Math.pow(policy.multiplier, attempt),
  );
  const jit = raw * policy.jitter * (2 * Math.random() - 1);
  return Math.max(0, raw + jit);
}

export async function withRetry<T>(
  fn: () => Promise<T>,
  options: {
    policy?: RetryPolicy;
    retryable?: (exc: unknown) => boolean;
  } = {},
): Promise<T> {
  const policy = options.policy ?? defaultRetryPolicy;
  const retryable = options.retryable ?? (() => true);
  const sleep = policy.sleep ?? defaultSleep;
  let lastError: unknown;
  for (let attempt = 0; attempt < policy.maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (exc) {
      lastError = exc;
      if (attempt === policy.maxAttempts - 1 || !retryable(exc)) {
        throw exc;
      }
      await sleep(delayForAttempt(policy, attempt));
    }
  }
  throw lastError;
}

File: packages/python/src/agent_capture/exporter/retry.py
"""Exponential backoff retry helper.

Used by :class:`agent_capture.exporter.http.HTTPExporter` for transient
failures (5xx, network errors). Never used for permanent failures (4xx,
malformed payload) — those drop after the first attempt with a loud
safelog entry.
"""

from __future__ import annotations

import random
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import TypeVar

T = TypeVar("T")


@dataclass(frozen=True)
class RetryPolicy:
    """Exponential backoff schedule.

    Sleep between attempt N and N+1 is::

        min(max_delay_s, base_delay_s * (multiplier ** N)) * (1 + jitter * U)

    where ``U`` is uniform in ``[-1, 1]``. Capped at ``max_delay_s``.
    """

    max_attempts: int = 5
    base_delay_s: float = 0.5
    max_delay_s: float = 30.0
    multiplier: float = 2.0
    jitter: float = 0.25
    sleep: Callable[[float], None] = time.sleep

    def delay(self, attempt: int) -> float:
        raw = min(self.max_delay_s, self.base_delay_s * (self.multiplier**attempt))
        jit = raw * self.jitter * (2 * random.random() - 1)
        return max(0.0, raw + jit)


def with_retry(
    fn: Callable[[], T],
    *,
    policy: RetryPolicy = RetryPolicy(),
    retryable: Callable[[BaseException], bool] = lambda _exc: True,
) -> T:
    """Call ``fn`` with exponential backoff. Re-raises the last error.

    Args:
        fn: Zero-arg callable that performs the operation.
        policy: Backoff schedule.
        retryable: Predicate determining whether an exception should
            trigger another attempt. Default: retry on everything.
            Pass a stricter predicate to bypass retries for permanent
            failures (4xx responses, schema errors, etc.).
    """
    last: BaseException | None = None
    for attempt in range(policy.max_attempts):
        try:
            return fn()
        except BaseException as exc:
            last = exc
            if attempt == policy.max_attempts - 1 or not retryable(exc):
                raise
            policy.sleep(policy.delay(attempt))
    assert last is not None  # unreachable
    raise last
